Add upgrade guide for `auto_new_epoch`

kmaehashi opened the pull request:
Add upgrade guide for ``auto_new_epoch`` introduced in #4608.

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
+Changed Behavior of ``Optimizer.new_epoch``
+-------------------------------------------
+
+This change only affects to a minor number of users (who call ``new_epoch()`` method of Optimizers while using a trainer, or who implement their own Updater class).
```
* "This change should affect only a minority of users"
* ... ```:meth:`Optimizer.new_epoch` ``` method manually
* ```:class:`Updater` ``` class

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
+
+This change only affects to a minor number of users (who call ``new_epoch()`` method of Optimizers while using a trainer, or who implement their own Updater class).
+
+Optimizers provide ``new_epoch()`` method, which can be used to change optimizer behavior depending on the current epoch number.
```
optimizer behavior > the behavior of the optimizer

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
+
+Optimizers provide ``new_epoch()`` method, which can be used to change optimizer behavior depending on the current epoch number.
+Prior to Chainer v5, this method was expected to be called by users.
+In Chainer v5, Updaters call ``new_epoch()`` method of Optimizers automatically.
```
* Updaters -> updaters (lower case)
* new_epoch() method of Optimizers -> ```:meth:`Optimizer.new_epoch` ```

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
+Optimizers provide ``new_epoch()`` method, which can be used to change optimizer behavior depending on the current epoch number.
+Prior to Chainer v5, this method was expected to be called by users.
+In Chainer v5, Updaters call ``new_epoch()`` method of Optimizers automatically.
+If you are calling ``new_epoch()`` method manually while using a trainer (Updater), you may need any of the following fixes:
```
(Updater) -> (or an updater)

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
+In Chainer v5, Updaters call ``new_epoch()`` method of Optimizers automatically.
+If you are calling ``new_epoch()`` method manually while using a trainer (Updater), you may need any of the following fixes:
+
+* Pass ``auto_new_epoch=False`` to the constructor of the Updater to stop ``new_epoch()`` being called automatically by Updaters.
```
* "Pass ... to `Updater.__init__` to stop..." (with a link to `Updater` class)
* ... from being called ...
* Updaters -> the updater

kmaehashi replied:
> "Pass ... to Updater.__init__ to stop..." (with a link to Updater class)

As `__init__` is not a part of Updater interface, I added link to `StandardUpdater` as an example.

niboshi replied:
"constructor" is not a correct term in Python.

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
+* Avoid calling ``new_epoch()`` method manually.
+
+If you are implementing your own Updater class, you may need to update your code to automatically call ``new_epoch()``.
+See `#4608 <https://github.com/chainer/chainer/pull/4608>`__ for the detailed discussions.
```
"... for a more detailed discussion" is more natural.
Btw, actually #4608 does not include any discussion. Is it worth mentioning?

kmaehashi replied:
There're no discussion, but updater implementors may find it useful to understand how to fix their own updaters.
Changed to:
> See `#4608 <https://github.com/chainer/chainer/pull/4608>`__ for more details.

niboshi replied:
I think the information in that PR is (and should be) written in this documentation.

kmaehashi replied:
I assume not many people implements own Updaters, so it's rather better to keep it simple.

kmaehashi commented:
Thanks for review! I fixed accordingly.

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -8,6 +8,26 @@ This is a list of changes introduced in each release that users should be aware
+* Pass ``auto_new_epoch=False`` to the constructor of the updater (e.g., :class:`~chainer.training.updaters.StandardUpdater`) to stop :meth:`~chainer.Optimizer.new_epoch` from being called automatically by the updater.
+* Avoid calling :meth:`~chainer.Optimizer.new_epoch` method manually.
+
+If you are implementing your own Updater class, you may need to update your code to automatically call :meth:`~chainer.Optimizer.new_epoch`.
```
Updater -> updater (lower case) or ```:class:`~chainer.training.Updater` ```

niboshi commented:
Also added some comments.

kmaehashi commented:
Fixed accordingly.
As discussed offline, I changed the sentence from:

> See `#4608 <https://github.com/chainer/chainer/pull/4608>`__ for the detailed discussions.

to:

> (you can refer to the changes introduced in `#4608 <https://github.com/chainer/chainer/pull/4608>`__ to understand how to fix your updater).

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -8,6 +8,25 @@ This is a list of changes introduced in each release that users should be aware
+Chainer v5
+==========
+
+Updaters Call ``Optimizer.new_epoch``
```
How about "Updators Automatically Call ..."?

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -8,6 +8,25 @@ This is a list of changes introduced in each release that users should be aware
+Updaters Call ``Optimizer.new_epoch``
+-------------------------------------
+
+This change should affect only a minority of users (who call :meth:`chainer.Optimizer.new_epoch` while using a trainer, or who implement their own :class:`~chainer.training.Updater` class).
```
Remove "chainer." prefix?

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -8,6 +8,25 @@ This is a list of changes introduced in each release that users should be aware
+
+Optimizers provide :meth:`~chainer.Optimizer.new_epoch` method, which can be used to change the behavior of optimizers depending on the current epoch number.
+Prior to Chainer v5, this method was expected to be called by users.
+In Chainer v5, updaters has been changed to call :meth:`~chainer.Optimizer.new_epoch` automatically.
```
updaters have been

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -8,6 +8,25 @@ This is a list of changes introduced in each release that users should be aware
+Optimizers provide :meth:`~chainer.Optimizer.new_epoch` method, which can be used to change the behavior of optimizers depending on the current epoch number.
+Prior to Chainer v5, this method was expected to be called by users.
+In Chainer v5, updaters has been changed to call :meth:`~chainer.Optimizer.new_epoch` automatically.
+If you are calling :meth:`~chainer.Optimizer.new_epoch` method manually while using a trainer (or an updater), you may need any of the following fixes:
```
"If you have been calling"?

niboshi commented on docs/source/upgrade.rst:
```diff
@@ -8,6 +8,25 @@ This is a list of changes introduced in each release that users should be aware
+* Pass ``auto_new_epoch=False`` to the constructor of the updater (e.g., :class:`~chainer.training.updaters.StandardUpdater`) to stop :meth:`~chainer.Optimizer.new_epoch` from being called automatically by the updater.
+* Avoid calling :meth:`~chainer.Optimizer.new_epoch` method manually.
+
+If you are implementing your own :class:`~chainer.training.Updater` class, you may need to update your code to automatically call :meth:`~chainer.Optimizer.new_epoch` (you can refer to the changes introduced in `#4608 <https://github.com/chainer/chainer/pull/4608>`__ to understand how to fix your updater).
```
"If you implement"?

kmaehashi commented:
Fixed.

niboshi commented:
LGTM.
